### PR TITLE
etcd-operator: bump to v0.5.0

### PR DESCRIPTION
**Pull request by klausenbusk**
Per v0.5.0 the cluster object is defined via a Custom Resource
Definition(CRD) instead of a Third Party Resource(TPR).
This commit also change all the relevant logic from TPR to CRD.

### CODE REVIEW

**Comment by klausenbusk** [comment]
Pleas see: https://github.com/kubernetes-incubator/bootkube/issues/666#issuecomment-322198664

**Comment by rphillips** [comment]
on the Sprintf... should the parameters be path.Escaped?

**Comment by klausenbusk** [comment]
I have just tested this with `hack/quickstart/init-master.sh` and it spin up a master node with `etcd-operator` `v0.5.0` without any issue. `bootkube recover` is also working without any issues.

File: pkg/recovery/etcd.go [hunk 1]
```diff
@@ -153,28 +152,22 @@ func (s *etcdSelfhostedBackend) read(ctx context.Context) (*controlPlane, error)
...
-	d, err := s.getBytes(ctx, etcdTPRKey)
+	d, err := s.getBytes(ctx, etcdCRDKey)
 	if err != nil {
 		return nil, err
 	}
 
-	var tpr v1beta1.ThirdPartyResourceData
-	err = decode(s.decoder, d, &tpr)
+	var kubeetcd spec.EtcdCluster
+	err = json.Unmarshal(d, &kubeetcd)
```
Comment (klausenbusk) [inline pkg/recovery/etcd.go]: This is the only serious change and it seems to work. `migrate-etcd-cluster.json` end up with the `status` field and the json isn't pretty-printed, but I think that is the "expected" behavior.

Comment (hongchaodeng) [inline pkg/recovery/etcd.go]: Well.. It could be protobuf.

**Comment by hongchaodeng** [comment]
The change generally looks good to me.
The one concern is about the codec change. If it's fixed to work with protobuf that it should be fine.

File: pkg/recovery/etcd.go (hunk 1 above)
Comment (klausenbusk) [inline pkg/recovery/etcd.go]: So something like:
```
err = json.Unmarshal(d, &kubeetcd)
if err != nil
    err = proto.Unmarshal(d, &kubeetcd)
    if != nil {
        return nil, err
    }
```
How can I know if it is profobuf? I didn't knew that k8s could use protobuf.

Comment (hongchaodeng) [inline pkg/recovery/etcd.go]: Why not keep original code:

```
err = decode(s.decoder, d, &tpr)
```

What's the problem with that?

**Comment by klausenbusk** [comment]
> Why not keep original code:

Oh, I see how it works now, I didn't exactly understand that `decoder` stuff before now :)

I have just changed that line and pushed it.

**Comment by hongchaodeng** [comment]
Great!
The change looks good to me. Let me give a try before merging it.

**Comment by hongchaodeng** [comment]
coreosbot test this please

**Comment by klausenbusk** [comment]
Failed, sadly and I can't see the log, hmm..

**Comment by hongchaodeng** [comment]
coreosbot retest this please

File: pkg/recovery/etcd.go [hunk 2]
```diff
@@ -121,7 +120,7 @@ func (s *etcdBackend) list(ctx context.Context, key string, listObj runtime.Obje
 
 const (
 	assetPathRecoveryEtcd  = "recovery-etcd.yaml"
-	etcdTPRKey             = "ThirdPartyResourceData/etcd.coreos.com/clusters/kube-system/kube-etcd"
```
Comment (hongchaodeng) [inline pkg/recovery/etcd.go]: This doesn't look like the right key

File: e2e/etcdscale_test.go [hunk 3]
```diff
@@ -15,7 +15,7 @@ import (
 )
 
 const (
-	kubeEtcdTPRURI = "/apis/etcd.coreos.com/v1beta1/namespaces/kube-system/clusters/kube-etcd"
+	kubeEtcdCRDURI = "/apis/etcd.database.coreos.com/v1beta2/namespaces/kube-system/clusters/kube-etcd"
```
Comment (hongchaodeng) [inline e2e/etcdscale_test.go]: clusters -> etcdclusters

Comment (hongchaodeng) [inline e2e/etcdscale_test.go]: Or better:
```
/apis/${spec.SchemeGroupVersion.String()}/namespaces/kube-system/${spec.CRDResourcePlural}/...
```

File: pkg/recovery/etcd.go (hunk 2 above)
Comment (hongchaodeng) [inline pkg/recovery/etcd.go]: OK.. After trying it, this is right..

**Comment by klausenbusk** [comment]
> /apis/${spec.SchemeGroupVersion.String()}/namespaces/kube-system/${spec.CRDResourcePlural}/...

Can I do that in a `const`?

File: e2e/etcdscale_test.go (hunk 3 above)
Comment (klausenbusk) [inline e2e/etcdscale_test.go]: > clusters -> etcdclusters

Done..

**Comment by hongchaodeng** [comment]
> Can I do that in a const?

Any methods that works.

File: e2e/etcdscale_test.go [hunk 4]
```diff
@@ -83,40 +82,41 @@ func etcdScalePreCheck(requiredMasters int) error {
 // resizes self-hosted etcd and checks that the desired number of pods are in a running state
 func resizeSelfHostedEtcd(t *testing.T, size int) {
 	httpRestClient := client.ExtensionsV1beta1().RESTClient()
-	var tpr unstructured.Unstructured
+	kubeEtcdCRDURI := fmt.Sprintf("/apis/%s/namespaces/%s/%s/kube-etcd", spec.SchemeGroupVersion.String(), api.NamespaceSystem, spec.CRDResourcePlural)
```
Comment (klausenbusk) [inline e2e/etcdscale_test.go]: @hongchaodeng How does this look?

Comment (hongchaodeng) [inline e2e/etcdscale_test.go]: 👍 

Can you rebase master?

**Comment by klausenbusk** [comment]
> Can you rebase master?

Done..

**Comment by hongchaodeng** [comment]
coreosbot retest this please

**Comment by hongchaodeng** [comment]
```

[335204.030083] golang[5]: # github.com/kubernetes-incubator/bootkube/e2e
[335204.030464] golang[5]: ./etcdscale_test.go:85: undefined: spec in spec.SchemeGroupVersion
[335204.030653] golang[5]: ./etcdscale_test.go:85: undefined: api in api.NamespaceSystem
[335204.030819] golang[5]: ./etcdscale_test.go:85: undefined: spec in spec.CRDResourcePlural
```

**Comment by klausenbusk** [comment]
> [335204.030083] golang[5]: # github.com/kubernetes-incubator/bootkube/e2e [335204.030464] golang[5]: ./etcdscale_test.go:85: undefined: spec in spec.SchemeGroupVersion [335204.030653] golang[5]: ./etcdscale_test.go:85: undefined: api in api.NamespaceSystem [335204.030819] golang[5]: ./etcdscale_test.go:85: undefined: spec in spec.CRDResourcePlural

My bad, I think it should work now, but I'm not exactly sure how to run the test.

**Comment by hongchaodeng** [comment]
coreosbot retest this please

**Comment by klausenbusk** [comment]
CI is green.

**Comment by hongchaodeng** [comment]
@klausenbusk 
We would need to manually verify it due to lack of some tests here. Hopefully by next week.

cc @xiang90 @hasbro17

**Comment by aaronlevy** [comment]
What are the tests that are missing? Can we add them rather than manually testing?

**Comment by hongchaodeng** [comment]
> What are the tests that are missing? Can we add them rather than manually testing?

Tectonic Installer test. That shouldn't be added here. I am just adding more confidence. That wouldn't block this PR from merging though.

File: e2e/etcdscale_test.go [hunk 5]
```diff
@@ -131,34 +133,34 @@ func resizeSelfHostedEtcd(t *testing.T, size int) {
...
+			log.Printf("Failed to unmarshal CRD: %v\n", err)
 			return false, nil
 		}
 
 		// check status of members
-		status, ok := tpr.Object["status"].(map[string]interface{})
+		status, ok := crd.Object["status"].(map[string]interface{})
 		if !ok {
-			log.Println("Could not asset 'status' type from TPR")
+			log.Println("Could not asset 'status' type from CRD")
```
Comment (squat) [inline e2e/etcdscale_test.go]: typo: s/asset/assert/

Comment (klausenbusk) [inline e2e/etcdscale_test.go]: Fixed..

**Comment by hongchaodeng** [comment]
@klausenbusk 
Can you squash the third commit into the second commit?

**Comment by klausenbusk** [comment]
> Can you squash the third commit into the second commit?

Done

**Comment by hongchaodeng** [comment]
coreosbot retest this please

**Comment by hongchaodeng** [comment]
coreosbot retest this please

**Comment by squat** [comment]
I just verified bootstrapping is working with @hongchaodeng. LGTM

**Pull request merged**